docs/proposals/apiserver_watch.md

```diff
@@ -0,0 +1,132 @@
+## Abstract
+
+In the current system, all watch requests send to apiserver are in general
+redirected to etcd. This means that for every watch request to apiserver,
+apiserver opens a watch on etcd.
+
+The purpose of the proposal is to improve the overall performance of the system
+by solving the following problems:
+
+- having too many open watches on etcd
+- avoiding deserializing/converting the same objects multiple times in different
+watch results
+
+In the future, we would also like to add an indexing mechanism to the watch.
+Although Indexer is not part of this proposal, it is supposed to be compatible
+with it - in the future Indexer should be incorporated into the proposed new
+watch solution in apiserver without requiring any redesign.
+
+
+## High level design
+
+We are going to solve those problems by allowing many clients to watch the same
+storage in the apiserver, without being redirected to etcd.
+
+At the high level, apiserver will have a single watch open to etcd, watching all
+the objects (of a given type) without any filtering. The changes delivered from
+etcd will then be stored in a cache in apiserver. The introduced cache will have
+a limited capacity - whenever a new change comes from etcd and the cache is full,
+the oldest change will be removed to make place for the new change.
```

Comment:
> One more thing I would like to emphasize - the whole idea behind
introducing the "Resync" watch event is to avoid a lot of LIST request to
etcd at the same time, whenever "out of window" error happens for the
"internal" watch. If we think this will not be a problem, maybe it doesn't
make sense to do it then (but I'm still not convinced).

Having the longer watch window, and being able to tell watch clients that
the last observed change to a particular watch subset (services, pods,
endpoints) happened at X but there has been no update since then should
reduce our LIST needs dramatically.

Our current problem is:

1. Setup watch on /services at X
2. No changes to services in 1000 events
3. Connection is broken, client last saw a change at X
4. Client tries to watch X again, gets error, now has to LIST

If the "unified watcher" in the apiserver knows that no changes have been
made to services since X, the following is possible

1. Setup watch on /services at X
2. No changes to services in 1000 events
3. Connection is broken, client last saw a change at X
4. Client tries to watch X again
5. Server knows that no services changes were made since X, server can
resume watch

Then there is no list and so Resync isn't necessary.


On Wed, Jul 15, 2015 at 4:17 PM, Daniel Smith <notifications@github.com>
wrote:

> In docs/proposals/apiserver_watch.md
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10679#discussion_r34723733>
> :
>
> > +In the future, we would also like to add an indexing mechanism to the watch.
> > +Although Indexer is not part of this proposal, it is supposed to be compatible
> > +with it - in the future Indexer should be incorporated into the proposed new
> > +watch solution in apiserver without requiring any redesign.
> > +
> > +
> > +## High level design
> > +
> > +We are going to solve those problems by allowing many clients to watch the same
> > +storage in the apiserver, without being redirected to etcd.
> > +
> > +At the high level, apiserver will have a single watch open to etcd, watching all
> > +the objects (of a given type) without any filtering. The changes delivered from
> > +etcd will then be stored in a cache in apiserver. The introduced cache will have
> > +a limited capacity - whenever a new change comes from etcd and the cache is full,
> > +the oldest change will be removed to make place for the new change.
>
> It has to be oldest, NOT least-used. Otherwise it won't be an effective
> replacement for etcd's watch window.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/10679/files#r34723733>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift